patches/api/0389-Add-extra-raytracing-api.patch

```diff
@@ -0,0 +1,415 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: PrimordialMoros <mail@moros.me>
+Date: Fri, 29 Jul 2022 16:04:32 +0300
+Subject: [PATCH] Add extra raytracing api
+
+Adds the option to ignore specific blocks (based on their position) when performing a raytrace.
+
+diff --git a/src/main/java/io/papermc/paper/raytrace/RayTraceContext.java b/src/main/java/io/papermc/paper/raytrace/RayTraceContext.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..0b3a490df1e114f81fcedc04343d4b105cb498fb
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/raytrace/RayTraceContext.java
+@@ -0,0 +1,320 @@
++package io.papermc.paper.raytrace;
++
++import com.google.common.base.Preconditions;
++import org.bukkit.FluidCollisionMode;
++import org.bukkit.Location;
++import org.bukkit.World;
++import org.bukkit.block.Block;
++import org.bukkit.entity.Entity;
++import org.bukkit.util.RayTraceResult;
++import org.bukkit.util.Vector;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.Collection;
++import java.util.Set;
++import java.util.function.Predicate;
++
++/**
++ * Represents the immutable context for performing a raytrace.
++ * <p>Usage of the {@link Builder} is preferred over the super long
++ * {@link World#rayTrace(Location, Vector, double, FluidCollisionMode, boolean, double, Predicate, Collection)} API.
++ */
++public class RayTraceContext {
```

Comment:
Can this be a record?

Reply:
It could but you'd need to repeat validation for all the arguments and also override some methods due to vectors being mutable. Not sure if it would be worth it.